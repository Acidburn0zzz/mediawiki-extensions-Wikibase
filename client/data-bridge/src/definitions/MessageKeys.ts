enum MessageKeys {
	BRIDGE_DIALOG_TITLE = 'wikibase-client-data-bridge-dialog-title',
	SAVE_CHANGES = 'savechanges',
}

export default MessageKeys;
