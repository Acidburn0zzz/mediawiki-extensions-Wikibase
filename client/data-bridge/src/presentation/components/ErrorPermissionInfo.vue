<template>
	<div class="wb-ui-permission-info-box">
		<div class="wb-ui-permission-info-box__info__icon">
			<div class="wb-ui-permission-info-box__header" v-html="messageHeader" />
		</div>
		<div class="wb-ui-permission-info-box__body">
			<div
				:class="[ infoIsExpanded ?
					'wb-ui-permission-info-box__icon--collapsed' :
					'wb-ui-permission-info-box__icon--expanded' ]"
				@click="toggleInfo"
			>
				<a class="wb-ui-permission-info-box__body__title">
					{{ this.$messages.get( this.$messages.KEYS.PERMISSIONS_MORE_INFO ) }}
				</a>
			</div>
			<div v-if="infoIsExpanded" v-html="messageBody" />
		</div>
	</div>
</template>

<script lang="ts">
import {
	Prop,
	Vue,
} from 'vue-property-decorator';
import Component from 'vue-class-component';

@Component
export default class ErrorPermissionInfo extends Vue {
	public infoIsExpanded = false;

	@Prop( { required: true } )
	private readonly messageHeader!: string;
	@Prop( { required: true } )
	private readonly messageBody!: string;

	private toggleInfo(): void {
		this.infoIsExpanded = !this.infoIsExpanded;
	}
}
</script>

<style lang="scss">
.wb-ui-permission-info-box {
	&__header {
		margin-left: 2em;
	}

	&__info__icon,
	&__icon--collapsed,
	&__icon--expanded {
		line-height: $size-icon;
		box-sizing: border-box;
		text-align: left;
		position: relative;
	}

	&__body {
		padding-left: 2em;

		&__title {
			margin-left: 2em;
		}
	}

	&__info__icon:before {
		background-image: $svg-info;
		background-repeat: no-repeat;
		background-size: contain;
		min-width: $min-size-icon;
		min-height: $min-size-icon;
		width: $size-icon;
		height: 100%;
		top: 0;
		position: absolute;
		content: '';
	}

	&__icon--collapsed:before {
		background-image: $svg-collapse;
		background-repeat: no-repeat;
		background-size: contain;
		min-width: $min-size-icon;
		min-height: $min-size-icon;
		width: $size-icon;
		height: 100%;
		top: 0;
		position: absolute;
		content: '';
	}

	&__icon--expanded:before {
		background-image: $svg-expand;
		background-repeat: no-repeat;
		background-size: contain;
		min-width: $min-size-icon;
		min-height: $min-size-icon;
		width: $size-icon;
		height: 100%;
		top: 0;
		position: absolute;
		content: '';
	}
}
</style>
