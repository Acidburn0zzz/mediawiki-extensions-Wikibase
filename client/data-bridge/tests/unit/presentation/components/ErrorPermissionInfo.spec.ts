import ErrorPermissionInfo from '@/presentation/components/ErrorPermissionInfo.vue';
import { shallowMount } from '@vue/test-utils';
import MessageKeys from '@/definitions/MessageKeys';

const TOGGLE_SELECTOR = '.wb-ui-permission-info-box__body div:nth-child(1)';
const BODY_SELECTOR = '.wb-ui-permission-info-box__body div:nth-child(2)';

describe( 'ErrorPermissionInfo', () => {
	it( 'matches the snapshot in closed state', () => {
		const messageHeader = 'header';
		const messageBody = 'body';
		const expandedByDefault = false;
		const wrapper = shallowMount( ErrorPermissionInfo, {
			propsData: { messageHeader, messageBody, expandedByDefault },
		} );
		expect( wrapper.element ).toMatchSnapshot();
	} );

	it( 'matches the snapshot in opened state', () => {
		const messageHeader = 'header';
		const messageBody = 'body';
		const expandedByDefault = true;
		const wrapper = shallowMount( ErrorPermissionInfo, {
			propsData: { messageHeader, messageBody, expandedByDefault },
		} );
		expect( wrapper.element ).toMatchSnapshot();
	} );

	it( 'uses correct message for toggle', () => {
		const toggleText = 'mytoggle';
		const messageGet = jest.fn().mockReturnValue( toggleText );
		const wrapper = shallowMount( ErrorPermissionInfo, {
			propsData: { messageHeader: 'header', messageBody: 'body' },
			mocks: {
				$messages: {
					KEYS: MessageKeys,
					get: messageGet,
				},
			},
		} );
		expect( messageGet ).toHaveBeenCalledTimes( 1 );
		expect( messageGet ).toHaveBeenCalledWith( MessageKeys.PERMISSIONS_MORE_INFO );
		expect( wrapper.find( TOGGLE_SELECTOR ).text() )
			.toBe( toggleText );
	} );

	it( 'Link toggles body visibility and link class', async () => {
		const expandedByDefault = false;
		const wrapper = shallowMount( ErrorPermissionInfo, {
			propsData: { messageHeader: 'header', messageBody: 'body', expandedByDefault },
		} );

		expect( wrapper.find( BODY_SELECTOR ).exists() )
			.toBeFalsy();
		wrapper.find( TOGGLE_SELECTOR ).trigger( 'click' );
		expect( wrapper.find( TOGGLE_SELECTOR ).classes() )
			.toContain( 'wb-ui-permission-info-box__icon--collapsed' );
		expect( wrapper.find( BODY_SELECTOR ).exists() )
			.toBeTruthy();
		wrapper.find( TOGGLE_SELECTOR ).trigger( 'click' );
		expect( wrapper.find( TOGGLE_SELECTOR ).classes() )
			.toContain( 'wb-ui-permission-info-box__icon--expanded' );
		expect( wrapper.find( BODY_SELECTOR ).exists() )
			.toBeFalsy();
	} );

} );
