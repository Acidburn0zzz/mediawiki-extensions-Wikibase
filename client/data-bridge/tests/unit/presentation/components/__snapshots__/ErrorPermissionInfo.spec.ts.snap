// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorPermissionInfo matches the snapshot in closed state 1`] = `
<div
  class="wb-ui-permission-info-box"
>
  <div
    class="wb-ui-permission-info-box__info__icon"
  >
    <div
      class="wb-ui-permission-info-box__header"
    >
      header
    </div>
  </div>
   
  <div
    class="wb-ui-permission-info-box__body"
  >
    <div
      class="wb-ui-permission-info-box__icon--expanded"
    >
      <a
        class="wb-ui-permission-info-box__body__title"
      >
        
				⧼wikibase-client-data-bridge-permissions-error-info⧽
			
      </a>
    </div>
     
    <!---->
  </div>
</div>
`;

exports[`ErrorPermissionInfo matches the snapshot in opened state 1`] = `
<div
  class="wb-ui-permission-info-box"
>
  <div
    class="wb-ui-permission-info-box__info__icon"
  >
    <div
      class="wb-ui-permission-info-box__header"
    >
      header
    </div>
  </div>
   
  <div
    class="wb-ui-permission-info-box__body"
  >
    <div
      class="wb-ui-permission-info-box__icon--collapsed"
    >
      <a
        class="wb-ui-permission-info-box__body__title"
      >
        
				⧼wikibase-client-data-bridge-permissions-error-info⧽
			
      </a>
    </div>
     
    <div>
      body
    </div>
  </div>
</div>
`;
